### PR TITLE
Update grpc.h comments for grpc_call thread safety

### PR DESCRIPTION
**Pull request by jtattermusch**
I've spotted that comments in grpc.h are not up to date and still refer to the old API.
Please check for correctness, what I put only reflect how I think things have evolved.

### CODE REVIEW

**Comment by jtattermusch** [comment]
Ping?

File: include/grpc/grpc.h [hunk 1]
```diff
@@ -387,7 +386,8 @@ grpc_call *grpc_channel_create_registered_call(
    completion of type 'tag' to the completion queue bound to the call.
    The order of ops specified in the batch has no significance.
    Only one operation of each type can be active at once in any given
-   batch. */
+   batch. 
+   THREAD SAFETY: grpc_call_start_batch is thread-compatible */
```
Comment (ctiller) [inline include/grpc/grpc.h]: There's a more complicated story here.

I think what we end up wanting is:
At most one thread can be starting a batch with GRPC_OP_SEND_* operations, and at most one thread can be starting a batch with GRPC_OP_RECV_* operations.

Comment (jtattermusch) [inline include/grpc/grpc.h]: Do you mean I can only be sending from one thread and receiving from one (possibly different) thread? That would be very broken (at least for C#).

Comment (ctiller) [inline include/grpc/grpc.h]: It's totally fine to be doing so. It's actually completely threadsafe to
start each different op type from a separate thread currently, but that
won't always be the case.

Comment (jtattermusch) [inline include/grpc/grpc.h]: I'd prefer to know that what C# is doing will be supported in the future. ;-)
Could you rephrase the restriction so I understand what you've actually meant? I can't guarantee at all from which thread the op will be call. Because the original docs said the call methods thread compatible I only initiate batched from within a lock owned by a call object, which is my understanding of how "thread-compatible" api method should be used.

Comment (ctiller) [inline include/grpc/grpc.h]: So what I said is more lenient that what you said.

What C# is doing is fine. A possible refactoring of what you're doing would
be to have two locks, one for reading, one for writing - and this would
still work - reads are thread-compatible, and writes are thread-compatible.
If you were starting a batch that did both, you'd need to take both mutexes
of course.

Comment (jtattermusch) [inline include/grpc/grpc.h]: Ah, thanks for clarification. So what you meant is:
1. access call_start_batch needs to be synchronized if used from multiple threads 
2. rule 1 is valid for all batch operations with the exception of  read and write operations, that can be synchronized separately (as an optimization allowing fully independent reading vs writing)

Comment (ctiller) [inline include/grpc/grpc.h]: Right

**Comment by jtattermusch** [comment]
Updated the comment for grpc_call_start_batch. PTAL.

**Comment by ctiller** [comment]
Needs merging. LGTM.

**Comment by jtattermusch** [comment]
Should be fine to merge now.
